

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tools (tl) &mdash; SpatioloJI 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=2d01f2b6" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=2d01f2b6" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=01f34227"></script>
      <script src="../_static/doctools.js?v=9bcbadda"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
      <script src="../_static/copybutton.js?v=f281be69"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Utilities (utils)" href="utils.html" />
    <link rel="prev" title="Plotting (pl)" href="pl.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            SpatioloJI
              <img src="../_static/logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#dependencies">Dependencies</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../basic_concepts.html">SpatioloJI Object Structure</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../basic_concepts.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../basic_concepts.html#core-components">Core Components</a></li>
<li class="toctree-l2"><a class="reference internal" href="../basic_concepts.html#component-details">Component Details</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#adata-anndata-object">adata (AnnData Object)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#cell-meta-dataframe">cell_meta (DataFrame)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#polygons-dataframe">polygons (DataFrame)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#fov-positions-dataframe">fov_positions (DataFrame)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#images-dictionary">images (Dictionary)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../basic_concepts.html#integration-and-relationships">Integration and Relationships</a></li>
<li class="toctree-l2"><a class="reference internal" href="../basic_concepts.html#creating-a-spatioloji-object">Creating a SpatioloJI Object</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#from-individual-components">From Individual Components</a></li>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#from-files">From Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#from-pickle">From Pickle</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../basic_concepts.html#key-methods">Key Methods</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#data-access">Data Access</a></li>
<li class="toctree-l3"><a class="reference internal" href="../basic_concepts.html#custom-data">Custom Data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../basic_concepts.html#summary">Summary</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">CosMx</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/index.html#basic-tutorials">Basic Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/QC_filtering.html">Quality Control and Filtering for Raw Data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../tutorials/QC_filtering.html#Prepare-original-data">Prepare original data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../tutorials/QC_filtering.html#Instantiation-of-Spatioloji_qc-object">Instantiation of Spatioloji_qc object</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/Scanpy_SpatioloJI.html">Integrate Scanpy into Spatiology Workflow</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../tutorials/Scanpy_SpatioloJI.html#Run-basic-workflow-of-Scanpy">Run basic workflow of Scanpy</a></li>
<li class="toctree-l4"><a class="reference internal" href="../tutorials/Scanpy_SpatioloJI.html#Imputation">Imputation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../tutorials/Scanpy_SpatioloJI.html#Instantiate-SpatioloJI-object">Instantiate SpatioloJI object</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/Basic_Visualization.html">Basic Visualization for Polygons or Dots in Spatial Images/ Stitiched Images</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../tutorials/Basic_Visualization.html#Plot">Plot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/index.html#advanced-tutorials">Advanced Tutorials</a><ul class="simple">
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Examples</a><ul class="simple">
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="data.html">Data (data)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="data.html#spatial-object">Spatial Object</a><ul>
<li class="toctree-l4"><a class="reference internal" href="data.html#SpatioloJI.data.Spatial_Object.Spatioloji"><code class="docutils literal notranslate"><span class="pre">Spatioloji</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="data.html#SpatioloJI.data.Spatial_Object.Spatioloji_qc"><code class="docutils literal notranslate"><span class="pre">Spatioloji_qc</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="pl.html">Plotting (pl)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="pl.html#plot-spatial-image">Plot Spatial Image</a><ul>
<li class="toctree-l4"><a class="reference internal" href="pl.html#SpatioloJI.pl.Plot_Spatial_Image.stitch_fov_images"><code class="docutils literal notranslate"><span class="pre">stitch_fov_images()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="pl.html#SpatioloJI.pl.Plot_Spatial_Image.plot_global_polygon_by_features"><code class="docutils literal notranslate"><span class="pre">plot_global_polygon_by_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="pl.html#SpatioloJI.pl.Plot_Spatial_Image.plot_global_polygon_by_categorical"><code class="docutils literal notranslate"><span class="pre">plot_global_polygon_by_categorical()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="pl.html#SpatioloJI.pl.Plot_Spatial_Image.plot_global_dot_by_features"><code class="docutils literal notranslate"><span class="pre">plot_global_dot_by_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="pl.html#SpatioloJI.pl.Plot_Spatial_Image.plot_global_dot_by_categorical"><code class="docutils literal notranslate"><span class="pre">plot_global_dot_by_categorical()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="pl.html#SpatioloJI.pl.Plot_Spatial_Image.plot_local_polygon_by_categorical"><code class="docutils literal notranslate"><span class="pre">plot_local_polygon_by_categorical()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="pl.html#SpatioloJI.pl.Plot_Spatial_Image.plot_local_polygon_by_features"><code class="docutils literal notranslate"><span class="pre">plot_local_polygon_by_features()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="pl.html#SpatioloJI.pl.Plot_Spatial_Image.plot_local_dots_by_categorical"><code class="docutils literal notranslate"><span class="pre">plot_local_dots_by_categorical()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="pl.html#SpatioloJI.pl.Plot_Spatial_Image.plot_local_dots_by_features"><code class="docutils literal notranslate"><span class="pre">plot_local_dots_by_features()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Tools (tl)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#spatial-functions">Spatial Functions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.perform_neighbor_analysis"><code class="docutils literal notranslate"><span class="pre">perform_neighbor_analysis()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_nearest_neighbor_distances"><code class="docutils literal notranslate"><span class="pre">calculate_nearest_neighbor_distances()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_cell_density"><code class="docutils literal notranslate"><span class="pre">calculate_cell_density()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_ripleys_k"><code class="docutils literal notranslate"><span class="pre">calculate_ripleys_k()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_cross_k_function"><code class="docutils literal notranslate"><span class="pre">calculate_cross_k_function()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_j_function"><code class="docutils literal notranslate"><span class="pre">calculate_j_function()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_g_function"><code class="docutils literal notranslate"><span class="pre">calculate_g_function()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_pair_correlation_function"><code class="docutils literal notranslate"><span class="pre">calculate_pair_correlation_function()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_cell_type_correlation"><code class="docutils literal notranslate"><span class="pre">calculate_cell_type_correlation()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_colocation_quotient"><code class="docutils literal notranslate"><span class="pre">calculate_colocation_quotient()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_proximity_analysis"><code class="docutils literal notranslate"><span class="pre">calculate_proximity_analysis()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_morisita_index"><code class="docutils literal notranslate"><span class="pre">calculate_morisita_index()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_quadrat_variance"><code class="docutils literal notranslate"><span class="pre">calculate_quadrat_variance()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_spatial_entropy"><code class="docutils literal notranslate"><span class="pre">calculate_spatial_entropy()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_hotspot_analysis"><code class="docutils literal notranslate"><span class="pre">calculate_hotspot_analysis()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_spatial_autocorrelation"><code class="docutils literal notranslate"><span class="pre">calculate_spatial_autocorrelation()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_kernel_density"><code class="docutils literal notranslate"><span class="pre">calculate_kernel_density()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_spatial_heterogeneity"><code class="docutils literal notranslate"><span class="pre">calculate_spatial_heterogeneity()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_mark_correlation"><code class="docutils literal notranslate"><span class="pre">calculate_mark_correlation()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_spatial_context"><code class="docutils literal notranslate"><span class="pre">calculate_spatial_context()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_network_statistics"><code class="docutils literal notranslate"><span class="pre">calculate_network_statistics()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#SpatioloJI.tl.Spatial_function.calculate_gene_spatial_autocorrelation"><code class="docutils literal notranslate"><span class="pre">calculate_gene_spatial_autocorrelation()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="utils.html">Utilities (utils)</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Development</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../contributing/index.html">Contributing to SpatioloJI</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../contributing/index.html#code-of-conduct">Code of Conduct</a></li>
<li class="toctree-l2"><a class="reference internal" href="../contributing/index.html#ways-to-contribute">Ways to Contribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="../contributing/index.html#getting-started">Getting Started</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../contributing/index.html#setting-up-your-development-environment">Setting Up Your Development Environment</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../contributing/index.html#development-workflow">Development Workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../contributing/index.html#making-changes">Making Changes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../contributing/index.html#code-style">Code Style</a></li>
<li class="toctree-l3"><a class="reference internal" href="../contributing/index.html#submitting-changes">Submitting Changes</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../contributing/index.html#documentation">Documentation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../contributing/index.html#writing-docstrings">Writing Docstrings</a></li>
<li class="toctree-l3"><a class="reference internal" href="../contributing/index.html#tutorials-and-examples">Tutorials and Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../contributing/index.html#testing">Testing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../contributing/index.html#writing-tests">Writing Tests</a></li>
<li class="toctree-l3"><a class="reference internal" href="../contributing/index.html#running-tests">Running Tests</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../contributing/index.html#release-process">Release Process</a></li>
<li class="toctree-l2"><a class="reference internal" href="../contributing/index.html#getting-help">Getting Help</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../changelog/index.html">changelog</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../changelog/index.html#version-0-2-0-2025-05-20">Version 0.2.0 (2025-05-20)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../changelog/index.html#added">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="../changelog/index.html#fixed">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../changelog/index.html#changed">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="../changelog/index.html#deprecated">Deprecated</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../ecosystem/index.html">Ji Universe Ecosystem</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../ecosystem/index.html#core-packages">Core Packages</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../ecosystem/index.html#spatioloji-core-foundation">SpatioloJI: Core Foundation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../ecosystem/index.html#dimensionji-multidimensional-analysis">DimensionJI: Multidimensional Analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../ecosystem/index.html#communicoji-cell-cell-communication">CommunicoJI: Cell-Cell Communication</a></li>
<li class="toctree-l3"><a class="reference internal" href="../ecosystem/index.html#fusionji-image-transcriptome-integration">FusionJI: Image-Transcriptome Integration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../ecosystem/index.html#encycloji-method-repository">EncycloJI: Method Repository</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../ecosystem/index.html#resources">Resources</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">SpatioloJI</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content style-external-links">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API Reference</a></li>
      <li class="breadcrumb-item active">Tools (tl)</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/api/tl.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tools-tl">
<h1>Tools (tl)<a class="headerlink" href="#tools-tl" title="Link to this heading"></a></h1>
<section id="spatial-functions">
<span id="module-SpatioloJI.tl"></span><h2>Spatial Functions<a class="headerlink" href="#spatial-functions" title="Link to this heading"></a></h2>
<dl class="py function" id="module-SpatioloJI.tl.Spatial_function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.perform_neighbor_analysis">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">perform_neighbor_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">polygon_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_metadata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'./'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'neighbor_analysis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_plots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(10,</span> <span class="pre">8)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#perform_neighbor_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.perform_neighbor_analysis" title="Link to this definition"></a></dt>
<dd><p>Perform comprehensive neighbor distance analysis for cell polygons.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>polygon_file</strong> (<em>pd.DataFrame</em>) – DataFrame with cell polygon vertices. Must include ‘cell’, ‘x_local_px’, ‘y_local_px’ columns.</p></li>
<li><p><strong>cell_metadata</strong> (<em>pd.DataFrame</em>) – DataFrame with cell metadata. Must include ‘cell’ and cell_type_column.</p></li>
<li><p><strong>cell_type_column</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – Column in cell_metadata that defines cell types for analysis.</p></li>
<li><p><strong>distance_threshold</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>, </em><em>optional</em>) – Maximum distance between polygons to be considered neighbors, by default 0.0
(0.0 means polygons must be touching, positive values include nearby non-touching cells)</p></li>
<li><p><strong>save_dir</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>optional</em>) – Directory to save result files, by default “./”</p></li>
<li><p><strong>filename_prefix</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>optional</em>) – Prefix for output filenames, by default “neighbor_analysis”</p></li>
<li><p><strong>include_plots</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to generate and save plots, by default True</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>]</em><em>, </em><em>optional</em>) – Figure size for plots, by default (10, 8)</p></li>
<li><p><strong>dpi</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.13)"><em>int</em></a><em>, </em><em>optional</em>) – DPI for saved figures, by default 300</p></li>
<li><p><strong>return_data</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to return the analysis data, by default True</p></li>
<li><p><strong>verbose</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.13)"><em>bool</em></a><em>, </em><em>optional</em>) – Whether to display progress information, by default True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary containing analysis results if return_data is True</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_nearest_neighbor_distances">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_nearest_neighbor_distances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_polygons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_nearest_neighbor_distances"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_nearest_neighbor_distances" title="Link to this definition"></a></dt>
<dd><p>Calculate the nearest neighbor distance for each cell.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>use_polygons</strong> – Whether to use polygon boundaries (True) or cell centers (False) for distance calculation</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with cell IDs and their nearest neighbor distances in pixels</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_cell_density">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_cell_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_by_area</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_polygons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_cell_density"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_cell_density" title="Link to this definition"></a></dt>
<dd><p>Calculate the cell density/crowding for each cell by counting neighbors within a specified radius.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>radius</strong> – Radius in pixels to search for neighboring cells</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>normalize_by_area</strong> – Whether to normalize count by the search area (True) or return raw counts (False)</p></li>
<li><p><strong>use_polygons</strong> – Whether to use polygon boundaries (True) or cell centers (False) for density calculation</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with cell IDs and their density measurements</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_ripleys_k">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_ripleys_k</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_ripleys_k"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_ripleys_k" title="Link to this definition"></a></dt>
<dd><p>Calculate Ripley’s K function to analyze spatial point patterns with multithreaded permutations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>max_distance</strong> – Maximum distance to calculate K function</p></li>
<li><p><strong>num_distances</strong> – Number of distance points to evaluate (default: 20)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>permutations</strong> – Number of Monte Carlo simulations for confidence envelope (0 for none)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for permutations (-1 for all processors)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with distances, K values, L values (normalized K), and optionally confidence envelopes</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_cross_k_function">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_cross_k_function</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_distances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_correction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_cross_k_function"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_cross_k_function" title="Link to this definition"></a></dt>
<dd><p>Calculate Ripley’s Cross-K function to analyze spatial relationships between two different cell types.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information</p></li>
<li><p><strong>cell_type1</strong> – First cell type for cross-K analysis</p></li>
<li><p><strong>cell_type2</strong> – Second cell type for cross-K analysis</p></li>
<li><p><strong>max_distance</strong> – Maximum distance to calculate K function (in pixels)</p></li>
<li><p><strong>num_distances</strong> – Number of distance points to evaluate (default: 20)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>edge_correction</strong> – Whether to apply edge correction (True) or not (False)</p></li>
<li><p><strong>permutations</strong> – Number of Monte Carlo simulations for confidence envelope (0 for none)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for permutations (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with distances, Cross-K values, and confidence envelopes if requested</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_j_function">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_j_function</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_correction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">monte_carlo_points</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_j_function"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_j_function" title="Link to this definition"></a></dt>
<dd><p>Calculate Baddeley’s J-function, which combines G-function and F-function for spatial pattern analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>max_distance</strong> – Maximum distance to consider (in pixels)</p></li>
<li><p><strong>num_bins</strong> – Number of distance bins for the functions</p></li>
<li><p><strong>cell_type</strong> – Optional cell type to filter cells (None for all cells)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type is provided)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>edge_correction</strong> – Whether to apply edge correction (True) or not (False)</p></li>
<li><p><strong>monte_carlo_points</strong> – Number of random points to generate for F-function calculation</p></li>
<li><p><strong>permutations</strong> – Number of Monte Carlo simulations for confidence envelope (0 for none)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for permutations (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with distances and J-function values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_g_function">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_g_function</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_correction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_g_function"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_g_function" title="Link to this definition"></a></dt>
<dd><p>Calculate the G-function (nearest neighbor distance distribution function).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>max_distance</strong> – Maximum distance to consider (in pixels)</p></li>
<li><p><strong>num_bins</strong> – Number of distance bins for the G-function</p></li>
<li><p><strong>cell_type</strong> – Optional cell type to filter cells (None for all cells)</p></li>
<li><p><strong>reference_cell_type</strong> – Optional reference cell type for cross-type G-function (None for same type)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type is provided)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>edge_correction</strong> – Whether to apply edge correction (True) or not (False)</p></li>
<li><p><strong>permutations</strong> – Number of Monte Carlo simulations for confidence envelope (0 for none)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for permutations (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with distances and G-function values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_pair_correlation_function">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_pair_correlation_function</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_correction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">monte_carlo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_pair_correlation_function"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_pair_correlation_function" title="Link to this definition"></a></dt>
<dd><p>Calculate the Pair Correlation Function (PCF) to analyze spatial relationships between cells at different distances.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>max_distance</strong> – Maximum distance to consider for the PCF (in pixels)</p></li>
<li><p><strong>num_bins</strong> – Number of distance bins to use</p></li>
<li><p><strong>cell_type1</strong> – Optional first cell type for type-specific PCF (None for all cells)</p></li>
<li><p><strong>cell_type2</strong> – Optional second cell type for type-specific PCF (None for all cells)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type1/2 is provided)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>edge_correction</strong> – Whether to apply edge correction (True) or not (False)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs (-1 for all processors)</p></li>
<li><p><strong>monte_carlo</strong> – Number of Monte Carlo simulations for confidence envelope (0 for none)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with distances and PCF values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_cell_type_correlation">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_cell_type_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_cell_type_correlation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_cell_type_correlation" title="Link to this definition"></a></dt>
<dd><p>Calculate spatial correlation between different cell types as a function of distance,
using multithreading for performance optimization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information</p></li>
<li><p><strong>max_distance</strong> – Maximum distance to consider for correlation analysis (in pixels)</p></li>
<li><p><strong>distance_bins</strong> – Number of distance bins to divide the analysis into</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs (-1 for all processors)</p></li>
<li><p><strong>batch_size</strong> – Number of cells to process in each parallel batch</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>pair_counts: DataFrame with counts of cell type pairs in each distance bin</p></li>
<li><p>correlation_matrix: Dict of correlation matrices for each distance bin</p></li>
<li><p>distance_bins: Array of distance bin edges</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict containing</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_colocation_quotient">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_colocation_quotient</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bootstrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">confidence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.95</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_colocation_quotient"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_colocation_quotient" title="Link to this definition"></a></dt>
<dd><p>Calculate the Colocation Quotient (CLQ) to measure spatial relationships between different cell types.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information</p></li>
<li><p><strong>distance_threshold</strong> – Maximum distance (in pixels) to consider cells as co-located</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>normalize</strong> – Whether to normalize the CLQ by the overall cell type frequencies</p></li>
<li><p><strong>bootstrap</strong> – Number of bootstrap iterations for confidence interval (0 for no bootstrapping)</p></li>
<li><p><strong>confidence</strong> – Confidence level for bootstrap interval (default: 0.95)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for bootstrap calculation (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with CLQ values for each pair of cell types</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_proximity_analysis">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_proximity_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_cell_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_cell_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_polygons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_correction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_proximity_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_proximity_analysis" title="Link to this definition"></a></dt>
<dd><p>Calculate proximity analysis to quantify spatial relationships between specific cell types.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>reference_cell_type</strong> – Cell type to use as reference points</p></li>
<li><p><strong>target_cell_types</strong> – List of cell types to analyze proximity to (None for all cell types)</p></li>
<li><p><strong>max_distance</strong> – Maximum distance to consider (in pixels) (None for auto-determination)</p></li>
<li><p><strong>distance_bins</strong> – Number of distance bins for proximity function</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>use_polygons</strong> – Whether to use cell polygons (True) or cell centers (False) for distance calculations</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information</p></li>
<li><p><strong>permutations</strong> – Number of Monte Carlo simulations for significance testing (0 for none)</p></li>
<li><p><strong>edge_correction</strong> – Whether to apply edge correction for boundary effects (True) or not (False)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for permutations (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict containing proximity statistics and spatial relationships between cell types</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_morisita_index">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_morisita_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bootstrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">confidence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.95</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_morisita_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_morisita_index" title="Link to this definition"></a></dt>
<dd><p>Calculate Morisita’s Index of Dispersion to analyze the spatial distribution pattern,
with multithreaded bootstrap confidence interval calculation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>grid_size</strong> – Size of the grid cells in pixels for quadrat counting</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>cell_type</strong> – Optional cell type to filter cells (None for all cells)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type is provided)</p></li>
<li><p><strong>bootstrap</strong> – Number of bootstrap iterations for confidence interval (0 for no bootstrapping)</p></li>
<li><p><strong>confidence</strong> – Confidence level for bootstrap interval (default: 0.95)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for bootstrap calculation (-1 for all processors)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict containing Morisita’s index and related statistics</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_quadrat_variance">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_quadrat_variance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_grid_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_grid_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_sizes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_quadrat_variance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_quadrat_variance" title="Link to this definition"></a></dt>
<dd><p>Calculate how variance of cell counts changes with quadrat size (quadrat variance analysis).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>min_grid_size</strong> – Minimum size of the grid cells in pixels</p></li>
<li><p><strong>max_grid_size</strong> – Maximum size of the grid cells in pixels</p></li>
<li><p><strong>num_sizes</strong> – Number of different grid sizes to test</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>cell_type</strong> – Optional cell type to filter cells (None for all cells)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type is provided)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs (-1 for all processors)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with grid sizes and corresponding variance metrics</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_spatial_entropy">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_spatial_entropy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entropy_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'shannon'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bootstrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_spatial_entropy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_spatial_entropy" title="Link to this definition"></a></dt>
<dd><p>Calculate spatial entropy to quantify the randomness in spatial distribution of cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>grid_size</strong> – Size of the grid cells in pixels for quadrat-based entropy calculation</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>cell_type</strong> – Optional cell type to filter cells (None for all cells)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type is provided)</p></li>
<li><p><strong>entropy_type</strong> – Type of entropy to calculate (‘shannon’, ‘renyi’, ‘tsallis’)</p></li>
<li><p><strong>normalization</strong> – Whether to normalize entropy by the maximum possible value</p></li>
<li><p><strong>bootstrap</strong> – Number of bootstrap iterations for confidence interval (0 for no bootstrapping)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for bootstrap calculation (-1 for all processors)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict containing entropy values and related statistics</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_hotspot_analysis">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_hotspot_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attribute_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p_value_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_hotspot_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_hotspot_analysis" title="Link to this definition"></a></dt>
<dd><p>Calculate Getis-Ord Gi* statistic to identify statistically significant spatial hot spots and cold spots.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>attribute_name</strong> – Name of the attribute/column in adata.obs to analyze for clustering</p></li>
<li><p><strong>distance_threshold</strong> – Maximum distance (in pixels) to consider neighbors</p></li>
<li><p><strong>p_value_threshold</strong> – Significance threshold for hot/cold spots (default: 0.05)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>permutations</strong> – Number of random permutations for significance testing</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with Gi* statistics and hot/cold spot classifications</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_spatial_autocorrelation">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_spatial_autocorrelation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attribute_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k_nearest</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_spatial_autocorrelation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_spatial_autocorrelation" title="Link to this definition"></a></dt>
<dd><p>Calculate spatial autocorrelation using Moran’s I statistic.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>attribute_name</strong> – Name of the attribute/column in adata.obs to analyze</p></li>
<li><p><strong>distance_threshold</strong> – Maximum distance (in pixels) to consider neighbors (None if using k_nearest)</p></li>
<li><p><strong>k_nearest</strong> – Number of nearest neighbors to use (None if using distance_threshold)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>local</strong> – Whether to calculate Local Moran’s I (True) or Global Moran’s I (False)</p></li>
<li><p><strong>permutations</strong> – Number of random permutations for significance testing</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Global Moran’s I result or DataFrame with Local Moran’s I values and quadrant classifications</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_kernel_density">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_kernel_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bandwidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kernel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gaussian'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">density_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_kernel_density"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_kernel_density" title="Link to this definition"></a></dt>
<dd><p>Calculate Kernel Density Estimation to visualize the spatial density of cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>bandwidth</strong> – Bandwidth for the kernel density estimator (None for automatic selection)</p></li>
<li><p><strong>grid_size</strong> – Number of grid points in each dimension for density estimation</p></li>
<li><p><strong>cell_type</strong> – Optional cell type to filter cells (None for all cells)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type is provided)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>kernel</strong> – Type of kernel to use (‘gaussian’, ‘epanechnikov’, ‘tophat’, etc.)</p></li>
<li><p><strong>normalize</strong> – Whether to normalize the density to sum to 1</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
<li><p><strong>density_threshold</strong> – Optional threshold to identify high-density regions (quantile between 0 and 1)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict containing KDE results, including density grid, coordinates, and metrics</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_spatial_heterogeneity">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_spatial_heterogeneity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attribute_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kernel_bandwidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'quadrat'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gradient_detection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_spatial_heterogeneity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_spatial_heterogeneity" title="Link to this definition"></a></dt>
<dd><p>Calculate spatial heterogeneity index to quantify and characterize spatial variation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>attribute_name</strong> – Column name in adata.obs containing attribute values (exclusive with gene_name)</p></li>
<li><p><strong>gene_name</strong> – Gene name to analyze expression heterogeneity (exclusive with attribute_name)</p></li>
<li><p><strong>grid_size</strong> – Size of grid cells in pixels for quadrat analysis (required for ‘quadrat’ method)</p></li>
<li><p><strong>kernel_bandwidth</strong> – Bandwidth for kernel density estimation (required for ‘kernel’ method)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>cell_type</strong> – Optional cell type to filter cells (None for all cells)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type is provided)</p></li>
<li><p><strong>method</strong> – Method for heterogeneity calculation (‘quadrat’, ‘kernel’, ‘nn’, or ‘variogram’)</p></li>
<li><p><strong>gradient_detection</strong> – Whether to detect and characterize spatial gradients (True) or not (False)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict containing heterogeneity metrics and spatial variation patterns</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_mark_correlation">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_mark_correlation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attribute_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mark_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'continuous'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_mark_correlation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_mark_correlation" title="Link to this definition"></a></dt>
<dd><p>Calculate the mark correlation function to analyze spatial correlation of cell attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>attribute_name</strong> – Name of the attribute/column in adata.obs to analyze</p></li>
<li><p><strong>max_distance</strong> – Maximum distance to consider (in pixels)</p></li>
<li><p><strong>num_bins</strong> – Number of distance bins for the correlation function</p></li>
<li><p><strong>cell_type</strong> – Optional cell type to filter cells (None for all cells)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type is provided)</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>mark_type</strong> – Type of attribute (‘continuous’ or ‘categorical’)</p></li>
<li><p><strong>permutations</strong> – Number of Monte Carlo simulations for confidence envelope (0 for none)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for permutations (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with distances and mark correlation values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_spatial_context">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_spatial_context</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_polygons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_by_distance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">context_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_spatial_context"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_spatial_context" title="Link to this definition"></a></dt>
<dd><p>Calculate the spatial context of cells based on their local neighborhood composition.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>distance_threshold</strong> – Maximum distance (in pixels) to consider cells as neighbors</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>use_polygons</strong> – Whether to use cell polygons for spatial relationships (True) or cell centers (False)</p></li>
<li><p><strong>weight_by_distance</strong> – Whether to weight neighbors by inverse distance (True) or use binary weights (False)</p></li>
<li><p><strong>context_types</strong> – Which cell types to include in context analysis (‘all’ or list of cell types)</p></li>
<li><p><strong>normalize</strong> – Whether to normalize context proportions (True) or use raw counts (False)</p></li>
<li><p><strong>min_neighbors</strong> – Minimum number of neighbors required for valid context analysis</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for context calculation (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with cell IDs and their spatial context profiles</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_network_statistics">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_network_statistics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k_nearest</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attribute_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_by_distance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">community_detection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centrality_measures</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_polygons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_overlap_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_network_statistics"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_network_statistics" title="Link to this definition"></a></dt>
<dd><p>Calculate network-based spatial statistics by constructing a cell interaction network,
with support for polygon-based cell geometries.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>distance_threshold</strong> – Maximum distance (in pixels) to consider cells as connected (None if using k_nearest)</p></li>
<li><p><strong>k_nearest</strong> – Number of nearest neighbors to connect for each cell (None if using distance_threshold)</p></li>
<li><p><strong>cell_type_column</strong> – Optional column name in adata.obs containing cell type information</p></li>
<li><p><strong>attribute_column</strong> – Optional column name in adata.obs containing attribute values for node analysis</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>directed</strong> – Whether to create a directed network (True) or undirected network (False)</p></li>
<li><p><strong>weight_by_distance</strong> – Whether to weight edges by inverse distance (True) or use binary weights (False)</p></li>
<li><p><strong>community_detection</strong> – Whether to perform community detection (True) or not (False)</p></li>
<li><p><strong>centrality_measures</strong> – Whether to calculate node centrality measures (True) or not (False)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for community detection and centrality calculation (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
<li><p><strong>use_polygons</strong> – Whether to use cell polygons for spatial relationships (True) or cell centers (False)</p></li>
<li><p><strong>edge_overlap_threshold</strong> – Minimum overlap ratio between polygon edges to create a connection (0.0-1.0)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dict containing network statistics, node metrics, and community structure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="SpatioloJI.tl.Spatial_function.calculate_gene_spatial_autocorrelation">
<span class="sig-prename descclassname"><span class="pre">SpatioloJI.tl.Spatial_function.</span></span><span class="sig-name descname"><span class="pre">calculate_gene_spatial_autocorrelation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spatioloji_obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_distance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distance_bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fov_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_global_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'moran'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/SpatioloJI/tl/Spatial_function.html#calculate_gene_spatial_autocorrelation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SpatioloJI.tl.Spatial_function.calculate_gene_spatial_autocorrelation" title="Link to this definition"></a></dt>
<dd><p>Calculate spatial autocorrelation of gene expression across the tissue.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>spatioloji_obj</strong> – A Spatioloji object</p></li>
<li><p><strong>genes</strong> – List of gene names or a single gene name to analyze</p></li>
<li><p><strong>max_distance</strong> – Maximum distance (in pixels) to consider for autocorrelation (None for auto-determination)</p></li>
<li><p><strong>distance_bins</strong> – Number of distance bins for correlogram analysis</p></li>
<li><p><strong>fov_id</strong> – Optional FOV ID to restrict analysis to a specific FOV</p></li>
<li><p><strong>use_global_coords</strong> – Whether to use global coordinates (True) or local coordinates (False)</p></li>
<li><p><strong>cell_type</strong> – Optional cell type to filter cells (None for all cells)</p></li>
<li><p><strong>cell_type_column</strong> – Column name in adata.obs containing cell type information (required if cell_type is provided)</p></li>
<li><p><strong>method</strong> – Method for spatial autocorrelation (‘moran’, ‘geary’, or ‘correlogram’)</p></li>
<li><p><strong>permutations</strong> – Number of Monte Carlo simulations for significance testing (0 for none)</p></li>
<li><p><strong>n_jobs</strong> – Number of parallel jobs for permutations (-1 for all processors)</p></li>
<li><p><strong>plot_result</strong> – Whether to return a plot configuration (True) or not (False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame with spatial autocorrelation statistics for each gene</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pl.html" class="btn btn-neutral float-left" title="Plotting (pl)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="utils.html" class="btn btn-neutral float-right" title="Utilities (utils)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, SpatioloJI Developers.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>